packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -49,56 +58,96 @@ export function extractLogSafeErrorProperties(error: any) {
+export interface FluidErrorAnnotations {
+    /** Telemetry props to log with the error */
+    props?: ITelemetryProperties;
+    /** fluidErrorCode to mention if error isn't already an IFluidErrorBase */
+    errorCodeIfNone?: string;
+}
+
+/** For backwards compatibility with pre-fluidErrorCode valid errors */
+function patchWithErrorCode(
+    legacyError: Omit<IFluidErrorBase, "fluidErrorCode">,
+    errorCode: string = "<error predates fluidErrorCode>",
+): asserts legacyError is IFluidErrorBase {
+    const patchMe: { fluidErrorCode?: string } = legacyError as any;
+    if (patchMe.fluidErrorCode === undefined) {
+        patchMe.fluidErrorCode = errorCode;
+    }
+}
 
 /**
- * Read-Write Logging Error.  Not exported.
- * This type alias includes addTelemetryProperties, and applies to objects even if not instanceof LoggingError\
+ * Normalize the given error yielding a valid Fluid Error
+ * @returns A valid Fluid Error with any provided annotations applied
+ * @param error - The error to normalize
+ * @param annotations - Annotations to apply to the normalized error
  */
-type RwLoggingError = LoggingError;
+export function normalizeError(
+    error: unknown,
+    annotations: FluidErrorAnnotations = {},
+): IFluidErrorBase {
+    // Back-compat, while IFluidErrorBase is rolled out
+    if (isValidLegacyError(error)) {
+        patchWithErrorCode(error, annotations.errorCodeIfNone);
+    }
+
+    if (isFluidError(error)) {
+        // We can simply annotate the error and return it
+        mixinTelemetryPropsWithFluidError(error, annotations.props);
+        return error;
+    }
+
+    // We have to construct a new fluid error, copying safe properties over
+    const { message, stack } = extractLogSafeErrorProperties(error);
+    const fluidError: IFluidErrorBase = {
+        errorType: "",
+        fluidErrorCode: annotations.errorCodeIfNone ?? "none",
+        message,
+        stack: stack ?? new Error("<<generated stack>>").stack,
+    };
+
+    mixinTelemetryPropsWithFluidError(fluidError, {
+        ...annotations.props,
+        untrustedOrigin: true, // This will let us filter to errors not originated by our own code
+        typeofError: typeof(error) === "object" ? undefined : typeof(error), // Only interesting for non-objects
+    });
+    return fluidError;
+}
+
+/** Mixin the telemetry props, along with errorType and fluidErrorCode */
+function mixinTelemetryPropsWithFluidError(
+    error: IFluidErrorBase,
+    props?: ITelemetryProperties,
+) {
+    const { errorType, fluidErrorCode } = error;
 
-/** Type guard for RwLoggingError.  Not exported. */
-const isRwLoggingError = (x: any): x is RwLoggingError =>
-    typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
+    // This is a back-compat move, so old loggers will include errorType and fluidErrorCode via prepareErrorObject
+    mixinTelemetryProps(error, { ...props, errorType, fluidErrorCode });
```

Comment:
I'd expect ...props to be the last in the list, such that if caller wants to overwrite errorType or fluidErrorCode, it can.

Reply:
Let's just tell people not to do that... 🤨